File: dotnet/src/MableBank.Application/TransferResultRecorder.cs
using MableBank.Domain;

namespace MableBank.Application;

/// <summary>Records transfer outcomes incrementally. Retains failures for reporting, not every success.</summary>
public sealed class TransferResultRecorder
{
    private readonly List<TransferResult> _failedResults = [];

    public IReadOnlyList<TransferResult> FailedResults => _failedResults;

    public int ProcessedCount { get; private set; }

    public int SucceededCount { get; private set; }

    public int FailedCount { get; private set; }

    public int SkippedCount { get; private set; }

    public TransferResult Record(TransferResult result)
    {
        ArgumentNullException.ThrowIfNull(result);
        ProcessedCount++;
        if (result.IsSuccess)
        {
            SucceededCount++;
        }
        else if (result.IsSkipped)
        {
            SkippedCount++;
        }
        else if (result.IsFailure)
        {
            FailedCount++;
            _failedResults.Add(result);
        }

        return result;
    }
}

File: nodejs-typescript/src/application/TransferResultRecorder.ts
import type { TransferResult } from '../domain/TransferResult.js';
import { isTransferFailure, isTransferSkipped, isTransferSuccess } from '../domain/TransferResult.js';

export type TransferFailureResult = Extract<TransferResult, { status: 'failure' }>;

/** Records transfer outcomes incrementally. Retains failures for reporting, not every success. */
export class TransferResultRecorder {
  readonly #failedResults: TransferFailureResult[] = [];
  processedCount = 0;
  succeededCount = 0;
  failedCount = 0;
  skippedCount = 0;

  get failedResults(): readonly TransferFailureResult[] {
    return this.#failedResults;
  }

  record(result: TransferResult): TransferResult {
    this.processedCount += 1;
    if (isTransferSuccess(result)) {
      this.succeededCount += 1;
    } else if (isTransferSkipped(result)) {
      this.skippedCount += 1;
    } else if (isTransferFailure(result)) {
      this.failedCount += 1;
      this.#failedResults.push(result);
    }
    return result;
  }
}
